proxy/__init__.py

```diff
@@ -55,15 +55,38 @@ def __init__(self, config: ProxyCacheConfig, repository: str, validation: bool =
             config.upstream_registry_hostname,
             config.upstream_registry_hostname,
         )
-        url = f"https://{hostname}"
-        if config.insecure:
-            url = f"http://{hostname}"
 
-        self.base_url = url
         self._session = requests.Session()
+
+        self.base_url = self._get_valid_upstream_registry(hostname)
         self._repo = repository
         self._authorize(self._credentials(), force_renewal=self._validation)
-        # flag used for validating Proxy cache config before saving to db
+        # flag used for validating Proxy cache config before saving to db    
+
+    def _get_valid_upstream_registry(self, hostname: str) -> str :
+        url = f"https://{hostname}"
+        try:
+            resp = self._session.get(f"{url}/v2")
+            if resp.ok:
+                return url
+        except requests.exceptions.SSLError:
+            if self._config.insecure:
+                # ignore verifying the SSL certificate
+                self._session.verify = False
+                resp = self._session.get(f"{url}/v2")
+                if resp.ok:
+                    return url
+        
+        # https + insecure failed, so fallback to http
+        url = f"http://{hostname}"
+        try:
+            resp = self._session.get(f"{url}/v2")
+        except requests.exceptions.SSLError as e:
```

Comment:
`requests` only raises a `SSLError` when SSL certificate validation fails[1] - there's no need to catch this exception here since the request at this point is using http, in which case there is no certificate to verify.

Here's the relevant bits from the docs:
> [...] and Requests will throw a SSLError if it’s unable to verify the certificate

[1] https://requests.readthedocs.io/en/latest/user/advanced/#ssl-cert-verification